File: Java/SubArraySum.java
import java.util.HashMap;

public class SubArraySum {
    public int subarraySum(int[] nums, int k) {
        int count = 0, prefixSum = 0;
        HashMap<Integer, Integer> sumMap = new HashMap<>();
        sumMap.put(0, 1);

        for (int num : nums) {
            prefixSum += num;
            if (sumMap.containsKey(prefixSum - k)) {
                count += sumMap.get(prefixSum - k);
            }
            sumMap.put(prefixSum, sumMap.getOrDefault(prefixSum, 0) + 1);
        }

        return count;
    }

    public static void main(String[] args) {
        SubArraySum sol = new SubArraySum();
        int[] nums = {1, 1, 1};
        int k = 2;
        System.out.println("Count: " + sol.subarraySum(nums, k));
    }
}

File: Python/SubArraySum.py
# Given an array of integers nums and an integer k, return the total number of continuous subarrays whose sum equals to k.


from typing import List
from collections import defaultdict

def subarraySum(nums: List[int], k: int) -> int:
    count = 0
    prefix_sum = 0
    prefix_map = defaultdict(int)
    prefix_map[0] = 1

    for num in nums:
        prefix_sum += num
        if prefix_sum - k in prefix_map:
            count += prefix_map[prefix_sum - k]
        prefix_map[prefix_sum] += 1

    return count

ls = [1, 2, 3]
k = 3
print(f"Total number of continuous subarrays whose sum equals to {k}: {subarraySum(ls, k)}")
